File: euler017.py
#! /usr/bin/python

onesText = { 1: "one",
             2: "two",
             3: "three",
             4: "four",
             5: "five",
             6: "six",
             7: "seven",
             8: "eight",
             9: "nine",
             10: "ten",
             11: "eleven",
             12: "twelve",
             13: "thirteen",
             14: "fourteen",
             15: "fifteen",
             16: "sixteen",
             17: "seventeen",
             18: "eighteen",
             19: "nineteen" }

tensText = { 2: "twenty",
             3: "thirty",
             4: "forty",
             5: "fifty",
             6: "sixty",
             7: "seventy",
             8: "eighty",
             9: "ninety"  }


def numberToText( n ) :
	text = ''
	if n == 1000 :
		text = "one thousand"
	else :
		hundreds = n // 100
		n %= 100	
		if hundreds :
			text += onesText[hundreds] + " hundred"
			if n :
				text += " and "
		tens = n // 10
		if tens >= 2 :
			n %= 10
			text += tensText[tens]
			if n :
				text += '-'
		if n :
			text += onesText[n]
	return text


total = 0
for i in xrange( 1, 1001 ) :
	result = numberToText( i )
	length = len( result.replace(' ','').replace('-','') )
	total = total + length

	print "%-35s : %5d" % (result, length)

print "-" * 45
print "%43d" % total

File: euler017.hs

onesText 1  = "one"
onesText 2  = "two"
onesText 3  = "three"
onesText 4  = "four"
onesText 5  = "five"
onesText 6  = "six"
onesText 7  = "seven"
onesText 8  = "eight"
onesText 9  = "nine"
onesText 10 = "ten"
onesText 11 = "eleven"
onesText 12 = "twelve"
onesText 13 = "thirteen"
onesText 14 = "fourteen"
onesText 15 = "fifteen"
onesText 16 = "sixteen"
onesText 17 = "seventeen"
onesText 18 = "eighteen"
onesText 19 = "nineteen"
onesText _  = ""

tensText 2 = "twenty"
tensText 3 = "thirty"
tensText 4 = "forty"
tensText 5 = "fifty"
tensText 6 = "sixty"
tensText 7 = "seventy"
tensText 8 = "eighty"
tensText 9 = "ninety"
tensText _ = ""

intToText :: Int -> String
intToText n = let (m,r1) = divMod n 1000
                  (c,r2) = divMod r1 100
                  (x,i) = if r2 >= 20 then divMod r2 10 else (0,r2)
                  mText = if m > 0 then onesText m ++ "thousand" else ""
                  cText = if c > 0 then onesText c ++ "hundred" else ""
                  andText = if (c > 0) && (r2 > 0) then "and" else ""
                  xText = tensText x
                  iText = onesText i
              in mText ++ cText ++ andText ++ xText ++ iText

main = print . sum . map (length . intToText) $ [1..1000]

